lib/compiler/src/compile_messages.erl

```diff
@@ -0,0 +1,214 @@
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+
+-module(compile_messages).
+
+-export([format_messages/4, list_errors/3]).
+
+-type pos() :: integer() | {integer(), integer()}.
+-type err_warn_info() :: tuple().
+-type message() :: {Where :: none | {File::string(), pos()}, Text :: iolist()}.
+
+-spec format_messages(File::string(), Prefix::string(), [err_warn_info()],
+                      Opts::[term()]) -> [message()].
+
+format_messages(F, P, [{none, Mod, E} | Es], Opts) ->
+    M = {none, io_lib:format("~ts: ~s~ts\n", [F, P, Mod:format_error(E)])},
+    [M | format_messages(F, P, Es, Opts)];
+format_messages(F, P, [{Loc, Mod, E} | Es], Opts) ->
+    StartLoc = erl_anno:location(Loc),
+    EndLoc =
+        case erl_anno:end_location(Loc) of
+            undefined -> StartLoc;
+            Loc2 -> Loc2
+        end,
+    Src = quote_source(F, StartLoc, EndLoc, Opts),
+    Msg = io_lib:format("~ts:~ts: ~s~ts\n~ts", [
+        F,
+        fmt_pos(StartLoc),
+        P,
+        Mod:format_error(E),
+        Src
+    ]),
+    Pos =
+        if
+            is_integer(StartLoc) -> {StartLoc, 0};
+            true -> StartLoc
+        end,
+    [{{F, Pos}, Msg} | format_messages(F, P, Es, Opts)];
+format_messages(_, _, [], _Opts) ->
+    [].
+
+-spec list_errors(File::string(), [err_warn_info()], Opts::[term()]) -> ok.
+
+list_errors(F, [{none, Mod, E} | Es], Opts) ->
+    io:fwrite("~ts: ~ts\n", [F, Mod:format_error(E)]),
+    list_errors(F, Es, Opts);
+list_errors(F, [{{{_, _} = StartLoc, {_, _} = EndLoc}, Mod, E} | Es], Opts) ->
+    %% this is the location format used in the type analysis pass
+    Src = quote_source(F, StartLoc, EndLoc, Opts),
+    io:fwrite("~ts:~ts: ~ts\n~ts", [F, fmt_pos(StartLoc), Mod:format_error(E), Src]),
+    list_errors(F, Es, Opts);
+list_errors(F, [{Loc, Mod, E} | Es], Opts) ->
+    StartLoc = erl_anno:location(Loc),
+    EndLoc =
+        case erl_anno:end_location(Loc) of
+            undefined -> StartLoc;
+            Loc2 -> Loc2
+        end,
+    list_errors(F, [{{StartLoc, EndLoc}, Mod, E} | Es], Opts);
+list_errors(_F, [], _Opts) ->
+    ok.
+
+fmt_pos({Line, Col}) ->
+    io_lib:format("~w:~w", [Line, Col]);
+fmt_pos(Line) ->
+    io_lib:format("~w", [Line]).
+
+quote_source(File, StartLoc, EndLoc, Opts) ->
+    case proplists:get_bool(brief, Opts) of
+        true -> "";
+        false -> quote_source_1(File, StartLoc, EndLoc)
+    end.
+
+quote_source_1(File, Line, Loc2) when is_integer(Line) ->
+    quote_source_1(File, {Line, 1}, Loc2);
+quote_source_1(File, Loc1, Line) when is_integer(Line) ->
+    quote_source_1(File, Loc1, {Line, -1});
+quote_source_1(File, {StartLine, StartCol}, {EndLine, EndCol}) ->
+    case file:read_file(File) of
```

Comment:
It looks like we are re-reading the source file for each error line. Is there a way to do fewer reads?

Reply:
Absolutely, this was a simplistic implementation. Considering OS file caches, it's probably not too bad as it is even for a hundred or more errors, but you could implement a caching system (which would have to handle include files as well, not just the main source file).

Reply:
Seems to be fast enough when compiling the compiler test suite (which has many warnings). In real code, there is probably very few warnings, as many projects use `-Werror` nowadays.